### DISCUSSION TITLE
Can't upload to ESP32-S3-DevKitC-1-N32R8V using Thonny

### DISCUSSION THREAD

**Discussion by ZacharyInDallas** [micropython/micropython#11041] [ESP32]
I bought this board:

ESP32-S3-DevKitC-1-N32R8V

WiFi Development Tools - 802.11 Gen-Purpose Dev Bd, embeds ESP32-S3-WROOM-2-N32R8V, 32 MB Octal flash & 8 MB Octal PSRAM, w/ Pin Header

I am using Thonny to upload my MicroPython code and data.
I can get a REP, which gives:

MicroPython v1.19.1 on 2022-06-18; ESP32S3 module with ESP32S3

When I try to upload a data file, it fails with:

PROBLEM IN THONNY'S BACK-END: Exception while handling 'upload' (thonny.plugins.micropython.mp_back.ManagementError: Command output was not empty).
See Thonny's backend.log for more info.
You may need to press "Stop/Restart" or hard-reset your MicroPython device and try again.

Process ended with exit code 1.0

If I try to upload main.py it fails with:

main.py (47803 bytes)

Error: Could not copy main.py to /main.py: Could not open file /main.py for writing, output:
Traceback (most recent call last):

  File "<stdin>", line 4, in <module>

OSError: [Errno 19] ENODEV

The first few lines of main.py are:

import configServer
import debug
import util
import machine

from micropython import const
import time

I tried the suggestions in the page below and it still does not work.  I tried both the release version and the daily build.

https://github.com/micropython/micropython/discussions/10206

**Comment by mattytrentini** [comment]
This device wasn't supported by the Espressif IDF until quite recently - if you're not using a recent version of the IDF to build with, you'll find that the filesystem does not mount and you'll get errors much like you're describing. 

I haven't gotten around to building this myself yet but I _believe_ it should be possible with IDF v4.4.3.

**Reply by ZacharyInDallas** [reply]
I'm not building it myself.  I'm using a pre-build binary.  I've tried both octal spiram versions and plain ones like GENERIC_S3. 
I tried the suggested pre-built binary in the web page below and it still does not work.  I tried both the release version and the daily build.

https://github.com/micropython/micropython/discussions/10206

**Reply by mattytrentini** [reply]
Sorry, I should have been clearer. I _believe_ that _none_ of the pre-built binaries will work. My understanding is that the build system that produces the pre-built binaries is currently using stable versions of the IDF and we need to be using more recent versions to provide support for 32MB flash. 

Until the build system is updated to use newer versions of the IDF I think you'll _have_ to build your own (using the GENERIC_S3_OCTAL board definition as a base but using a larger flash definition).

I do have one of these boards and am keen to get it working but it's not _quite_ at the top of my priority list. I'll try and get to it sooner rather than later!

**Reply by ZacharyInDallas** [reply]
I don't have to use the particular board that I bought.  I chose it because it seemed to be one of the fastest with lots of RAM and flash.  What ESP32 board can I buy that is fast, has lots of memory, and has a pre-built binary that will work?

**Reply by ZacharyInDallas** [reply]
Is there a way to be notified when a pre-built binary is available for this board that works?

**Comment by robert-hh** [comment]
It may be a problem of the interaction of your board with Thonny. Please try mpremote as the official tool to upload scripts, e.g.

`mpremote cp <filename> :`

`:` is the prefix for the board's file system. Use `mpremote --help` for  a list of options. You can use `mpremote ls` to show the files on the board.

**Reply by ZacharyInDallas** [reply]
Tried that and got a similar error:

cp main.py :
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 19] ENODEV

One problem I had doing the cp command is that I wanted to use com11 and there was something at com3.  It kept using com3 because it was the first USB device.  I finally got around that by connecting to com3 in one window and them doing the cp command in another terminal window.  Is there a better way of doing this?

**Reply by robert-hh** [reply]
That's strange, since `OSError: [Errno 19] ENODEV` is usually thrown when there is no file system. But a file system should be created at first boot. Anyhow, you can create one with:

```
import inisetup
vfs = inisetup.setup()
```

**Reply by robert-hh** [reply]
About com ports: You can tell mpremote which com port to use by e.g.:

mpremote connect com13 ls

**Reply by ZacharyInDallas** [reply]
If I do mpremote connect com11 ls
I get:
ls:
Does that indicate there is an empty file system, or could it be that there is no file system?

Is the:
import inisetup
vfs = inisetup.setup()

done in a REPL window for that microprocessor or do you do it in code you are uploading?

**Reply by robert-hh** [reply]
The latter in REPL.

**Reply by ZacharyInDallas** [reply]
Tried inisetup.setup:

MicroPython v1.19.1 on 2022-06-18; ESP32S3 module with ESP32S3

Type "help()" for more information.

>>> import inisetup
>>> vfs = inisetup.setup()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "inisetup.py", line 34, in setup
  File "inisetup.py", line 7, in check_bootsec
OSError: (-24579, 'ESP_ERR_FLASH_NOT_INITIALISED')
>>>

**Reply by robert-hh** [reply]
Then try a direct format:
```
import uos
from flashbdev import bdev
uos.VfsLfs2.mkfs(bdev)
vfs = uos.VfsLfs2(bdev)
uos.mount(vfs, "/")
```
If that fails, there is a problem with flash.

**Reply by ZacharyInDallas** [reply]
MicroPython v1.19.1 on 2022-06-18; ESP32S3 module with ESP32S3

Type "help()" for more information.

>>> import uos
>>> from flashbdev import bdev
>>> uos.VfsLfs2.mkfs(bdev)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: (-24579, 'ESP_ERR_FLASH_NOT_INITIALISED')
>>>

**Reply by robert-hh** [reply]
That does not look good, since mkfs() unconditionally creates the file system. The next thing I would try to completely wipe the flash using esptool.py. Then, MicroPython has to be reinstalled.

**Reply by robert-hh** [reply]
Reading back the comment of Matt, that will not work, since the octal flash is mot supported by the pre-built binaries. You have to build your own firmware based on esp-idf v4.4.3. Building a firmware is best (maybe only) done at a Linux system.

**Reply by robert-hh** [reply]
So I have built a firmware which might work: https://github.com/robert-hh/Shared-Stuff/blob/master/firmware_ESP32_GENERIC_C3_SPIRAM_OCT.bin
I cannot test it because I have no C3 board with octal memories. Maybe it helps.

**Reply by ZacharyInDallas** [reply]
I tried a plain pre-build binary like GENERIC_S3 which does not mention octal and that did not work.  Does that make a difference?

Also, the pre-build binary below claims to support octal spiram.  Do you think it supports octal spiram but not octal flash?
https://www.micropython.org/download/GENERIC_S3_SPIRAM_OCT/

**Reply by robert-hh** [reply]
Don't know. Just try it.ESP32 has the bootloader in ROM, so you cannot brick it.

**Reply by robert-hh** [reply]
The name of the firmware I linked was wrong. In fact is was the S3 build with the v4.4.3 esp-idf firmware. https://github.com/robert-hh/Shared-Stuff/blob/master/firmware_ESP32_GENERIC_S3_SPIRAM_OCT.bin

**Reply by dhalbert** [reply]
We got this board to work in CircuitPython. Here's our board-specific sdkconfig: https://github.com/adafruit/circuitpython/blob/main/ports/espressif/boards/espressif_esp32s3_devkitc_1_n32r8/sdkconfig. It may be slightly helpful.

**Reply by robert-hh** [reply]
It looks like the config for the ESP32_GENERIC_S3_SPIRAM_OCT build has the flag for octal spiram set, at a flash size of 8MB.

**Reply by ZacharyInDallas** [reply]
I don't have to use the particular board that I bought. I chose it because it seemed to be one of the fastest with lots of RAM and flash. What ESP32 board can I buy that is fast, has lots of memory, and has a pre-built binary that will work?

**Reply by Redika-ak** [reply]
this one https://micropython.org/download/ESP32_GENERIC_S3/

**Comment by cloudwithadam** [comment]
I have the same board but N8R8. The ESP32_GENERIC_S3_SPIRAM_OCT nightly build works for me.
